tools/graphiteindex/src/main/java/com/criteo/biggraphite/graphiteindex/MetricsIndex.java

```diff
@@ -0,0 +1,262 @@
+     */
+    public synchronized void forceMerge()
+    {
+        logger.debug("{} - Forcing merge of index segments", name);
+
+        try {
+            writer.forceMerge(1);
+            forceCommit();
+        } catch(IOException e) {
+            logger.error("{} - Cannot force merge of index segments", name, e);
+        }
+    }
+
+    public void insert(String path)
+    {
+        logger.trace("{} - Inserting '{}'", name, path);
+
+        Document doc = MetricPath.toDocument(path);
+
+        try {
+            writer.addDocument(doc);
+        } catch(IOException e) {
+            logger.error("{} - Cannot insert metric in index", name, e);
+        }
+    }
+
+    public void insert(String path, long offset)
+    {
+        logger.trace("{} - Inserting '{}' with offset {}", name, path);
+
+        Document doc = MetricPath.toDocument(path, offset);
+
+        try {
+            writer.addDocument(doc);
+        } catch(IOException e) {
+            logger.error("{} - Cannot insert metric in index", name, e);
+        }
+    }
+
+    public List<Long> searchOffsets(String pattern)
+    {
+        return search(pattern, MetricPath::getOffsetFromDocument);
+    }
+
+    public List<Pair<String, Long>> searchPaths(String pattern)
+    {
+        return search(
+            pattern,
+            doc -> Pair.of(
+                MetricPath.getPathFromDocument(doc),
+                MetricPath.getOffsetFromDocument(doc)
+            )
+        );
+    }
+
+    private <T> List<T> search(String pattern, Function<Document, T> handler)
+    {
+        BooleanQuery query = patternToQuery(pattern);
+        logger.trace("{} - Searching for '{}', generated query: {}", name, pattern, query);
+
+        ArrayList<T> results = new ArrayList<>();
+        Collector collector = new MetricsIndexCollector(
+            doc -> results.add(handler.apply(doc))
+        );
+
+        try {
+            IndexSearcher searcher = searcherMgr.acquire();
+            try {
+                // TODO(d.forest): we should probably use TimeLimitingCollector
+                searcher.search(query, collector);
+            } catch(IOException e) {
+                logger.error("{} - Cannot finish search query", name, e);
+            } finally {
+                searcherMgr.release(searcher);
+            }
+        } catch(IOException e) {
+            logger.error("{} - Cannot acquire index searcher", name, e);
+        }
+
+        return results;
+    }
+
+    /**
+     * Generates a Lucene query which matches the same metrics the given Graphite
+     * globbing pattern would.
+     */
+    private BooleanQuery patternToQuery(String pattern)
+    {
+        BooleanQuery.Builder queryBuilder = new BooleanQuery.Builder();
+
+        int length = MetricPath.iterateOnElements(
+            pattern,
+            (element, depth) -> {
+                // Wildcard-only fields do not filter anything
+                // TODO(d.forest): detect several wildcards?
```

Comment:
maybe add a todo to add support for globstar here

Reply:
correct, globstar and autocomplete are currently not implemented